flink-table/flink-sql-gateway/src/main/java/org/apache/flink/table/gateway/service/context/SessionContext.java

```diff
@@ -235,40 +236,79 @@ public static SessionContext create(
+        // Build session context and return
+        // --------------------------------------------------------------------------------------------------------------
+
+        SessionContext sessionContext =
+                new SessionContext(
+                        defaultContext,
+                        sessionId,
+                        environment.getSessionEndpointVersion(),
+                        configuration,
+                        userClassLoader,
+                        sessionState,
+                        new OperationManager(operationExecutorService));
+
+        // filter the default catalog out to avoid exception caused by repeated registration
+        String currentCatalog = sessionContext.sessionState.catalogManager.getCurrentCatalog();
+        environment
+                .getRegisteredCatalogs()
+                .forEach(
+                        (catalogName, catalog) -> {
+                            if (!catalogName.equals(currentCatalog)) {
+                                sessionContext.registerCatalog(catalogName, catalog);
+                            }
+                        });
+
+        environment.getRegisteredModules().forEach(sessionContext::registerModuleAtHead);
+
+        return sessionContext;
     }
 
     // ------------------------------------------------------------------------------------------------------------------
     // Helpers
     // ------------------------------------------------------------------------------------------------------------------
 
+    private static CatalogManager buildCatalogManager(
+            Configuration configuration,
+            URLClassLoader userClassLoader,
+            SessionEnvironment environment) {
+        CatalogManager.Builder builder =
+                CatalogManager.newBuilder()
+                        // Currently, the classloader is only used by DataTypeFactory.
+                        .classLoader(userClassLoader)
+                        .config(configuration);
+
+        // init default catalog
+        String defaultCatalogName;
+        Catalog defaultCatalog;
+        if (environment.getDefaultCatalog().isPresent()) {
+            defaultCatalogName = environment.getDefaultCatalog().get();
+            defaultCatalog = environment.getRegisteredCatalogs().get(defaultCatalogName);
+        } else {
+            EnvironmentSettings settings =
+                    EnvironmentSettings.newInstance().withConfiguration(configuration).build();
+            defaultCatalogName = settings.getBuiltInCatalogName();
+            defaultCatalog =
+                    new GenericInMemoryCatalog(
+                            defaultCatalogName, settings.getBuiltInDatabaseName());
+        }
+        defaultCatalog.open();
+
+        return builder.defaultCatalog(defaultCatalogName, defaultCatalog)
+                .currentDatabase(environment.getDefaultDatabase().orElse(null))
```

Comment:
After rethinking this problem, I think we should not set the current database explicitly. It is the catalog's attributes. We should set the default database when build the catalog.

Reply:
I removed 'CatalogManager.Builder#currentDatabase'.